From 6c62c91d27526d9059a5f1f676a9c790b85167dc Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Tue, 21 Jul 2026 18:37:03 -0400
Subject: [PATCH 211/268] winewayland: Avoid repositioning client surfaces when
 unnecessary.

See mahkoh(slash)jay(slash)issues(slash)1124#issuecomment-5039483751
---
 dlls/winewayland.drv/wayland_surface.c | 27 +++++++++++++++++++-------
 dlls/winewayland.drv/waylanddrv.h      |  1 +
 2 files changed, 21 insertions(+), 7 deletions(-)

diff --git a/dlls/winewayland.drv/wayland_surface.c b/dlls/winewayland.drv/wayland_surface.c
index e66bfc2c0f1..82d9b2b7143 100644
--- a/dlls/winewayland.drv/wayland_surface.c
+++ b/dlls/winewayland.drv/wayland_surface.c
@@ -943,7 +943,7 @@ static void wayland_surface_reconfigure_size(struct wayland_surface *surface,
  *
  * Reconfigures the subsurface covering the client area.
  */
-static void wayland_surface_reconfigure_client(struct wayland_surface *surface,
+static BOOL wayland_surface_reconfigure_client(struct wayland_surface *surface,
                                                struct wayland_client_surface *client,
                                                const RECT *client_rect)
 {
@@ -957,16 +957,18 @@ static void wayland_surface_reconfigure_client(struct wayland_surface *surface,
 
     TRACE("hwnd=%p rect=%s\n", surface->hwnd, wine_dbgstr_rect(&rect));
 
+    if (EqualRect(&rect, &client->viewport_rect)) return FALSE;
+    client->viewport_rect = rect;
+
     if (client->wl_subsurface)
-    {
         wl_subsurface_set_position(client->wl_subsurface, rect.left, rect.top);
-        wl_subsurface_place_above(client->wl_subsurface, surface->wl_surface);
-    }
 
     if (rect.left != rect.right && rect.top != rect.bottom)
         wp_viewport_set_destination(client->wp_viewport, rect.right - rect.left, rect.bottom - rect.top);
     else /* We can't have a 0x0 destination, use 1x1 instead. */
         wp_viewport_set_destination(client->wp_viewport, 1, 1);
+
+    return TRUE;
 }
 
 /**********************************************************************
@@ -1640,6 +1642,8 @@ struct wayland_client_surface *wayland_client_surface_create(HWND hwnd)
         }
     }
 
+    SetRect(&client->viewport_rect, -INT32_MAX, 0, -INT32_MAX, 0);
+
     return client;
 
 err:
@@ -1662,6 +1666,7 @@ void wayland_client_surface_attach(struct wayland_client_surface *client, HWND t
             client->wl_subsurface = NULL;
         }
 
+        SetRect(&client->viewport_rect, -INT32_MAX, 0, -INT32_MAX, 0);
         client->toplevel = 0;
         return;
     }
@@ -1682,6 +1687,12 @@ void wayland_client_surface_attach(struct wayland_client_surface *client, HWND t
                                             surface->wl_surface);
         if (!client->wl_subsurface) goto done;
 
+        /* If the parent wl_surface changes, then the subsurface will be reattached.
+         * So this place_above only need to be done once on attach. If two child windows
+         * are on top of each other then the composition behavior depends on the clip rect
+         * of each window rather than the Z-ordering of the child and parent. */
+        wl_subsurface_place_above(client->wl_subsurface, surface->wl_surface);
+
         /* Present contents independently of the parent surface. */
         wl_subsurface_set_desync(client->wl_subsurface);
 
@@ -1693,9 +1704,11 @@ void wayland_client_surface_attach(struct wayland_client_surface *client, HWND t
     NtUserGetClientRect(hwnd, &client_rect, NtUserGetWinMonitorDpi(hwnd, MDT_RAW_DPI));
     NtUserMapWindowPoints(hwnd, toplevel, (POINT *)&client_rect, 2, NtUserGetWinMonitorDpi(hwnd, MDT_RAW_DPI));
 
-    wayland_surface_reconfigure_client(surface, client, &client_rect);
-    /* Commit to apply subsurface positioning. */
-    wl_surface_commit(surface->wl_surface);
+    if (wayland_surface_reconfigure_client(surface, client, &client_rect))
+    {
+        /* Commit to apply subsurface positioning. */
+        wl_surface_commit(surface->wl_surface);
+    }
 
 done:
     wayland_win_data_release(toplevel_data);
diff --git a/dlls/winewayland.drv/waylanddrv.h b/dlls/winewayland.drv/waylanddrv.h
index aa2ccfe93c5..675c357ce35 100644
--- a/dlls/winewayland.drv/waylanddrv.h
+++ b/dlls/winewayland.drv/waylanddrv.h
@@ -362,6 +362,7 @@ struct wayland_client_surface
     struct client_surface client;
     HWND toplevel;
     HANDLE throttle;
+    RECT viewport_rect;
     struct wl_callback *wl_callback;
     struct wl_surface *wl_surface;
     struct wl_subsurface *wl_subsurface;
-- 
2.55.0

